### PR TITLE
Feature/permissions

### PR DESCRIPTION
**Pull request by ScottBTR**
### Description of Change ###
RequestPermissions:
- Request Calendar Read Permissions
- Request Calendar Write Permissions 
- NOTE: On iOS there is no separation of Read/Write for Calendar permissions so either call will request calendar permissions.

GetCalendarsAsync:
- Check/Request calendar read permissions
- Declare calendar instance/singleton
- Request List of iOS/Android platform specific calenders
- Cast to new shared List<DeviceCalendar> which obtains values for
* Id - Platform specific android generally holds number ids whilst iOS uses guids
* Name - Name of each calendar
* IsReadOnly - Can this calendar be edited
- Currently the sample app only displays the name attribute, but we could indicate read only access and ids at a later state.

### Bugs Fixed ###

- Related to issue #1, #2, #3 

Provide links to issues here. Ensure that a GitHub issue was created for your feature or bug fix before sending PR.

### API Changes ###

List all API changes here (or just put None), example:

```csharp
public static partial class Calendar {
  public static bool IsSupported();
  public static Task<IReadOnlyList<ICalendar>> GetCalendarsAsync();
  public static Task RequestCalendarReadAccess();
  public static Task RequestCalendarWriteAccess();
}
```

### Behavioral Changes ###

There is a new section in the samples list called Calendar:
- Contains 3 buttons:
  * Request Read Access - Prompts the user for calendar read access
  * Request Write Access - Prompts the user for calendar write access
  * Load Calendars - Prompts the user for calendar read access if not supplied, then fetches a list of calendars and provides a List View containing the name of each calendar found.
### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [X] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

File: Samples/Samples.iOS/Info.plist [hunk 1]
```diff
@@ -55,6 +55,10 @@
 	<string>1.0</string>
 	<key>NSLocationWhenInUseUsageDescription</key>
 	<string>Access to your location is required for cool things to happen!</string>
+    <key>NSCalendarsUsageDescription</key>
+    <string>Add calendar events</string>
+    <key>NSRemindersUsageDescription</key>
+    <string>Add calendar events</string>
```
Comment (nickrandolph) [inline Samples/Samples.iOS/Info.plist]: Let's update these strings to be more useful eg
View and create calendar events
View and create reminders

File: Samples/Samples/ViewModel/CalendarViewModel.cs [hunk 2]
```diff
@@ -1,15 +1,72 @@
...
 using System.Windows.Input;
 using Xamarin.Essentials;
 using Xamarin.Forms;
 
 namespace Samples.ViewModel
 {
-    class CalendarViewModel : BaseViewModel
+    public class CalendarViewModel : BaseViewModel
     {
+        // public ObservableCollection<Calendar> calendars;
```
Comment (nickrandolph) [inline Samples/Samples/ViewModel/CalendarViewModel.cs]: Remove commented code

File: Samples/Samples/ViewModel/CalendarViewModel.cs [hunk 3]
```diff
@@ -1,15 +1,72 @@
...
+
+        async void OnRequestCalendarWriteAccess()
+        {
+            try
+            {
+                await Calendar.RequestCalendarWriteAccess();
+            }
+            catch (PermissionException ex)
+            {
+                await DisplayAlertAsync($"Unable to request calendar read access: {ex.Message}");
```
Comment (nickrandolph) [inline Samples/Samples/ViewModel/CalendarViewModel.cs]: Change to "write" access (and in Exception handler below)

File: Tests/Calendar_Tests.cs [hunk 4]
```diff
@@ -8,6 +8,6 @@ public class Calendar_Tests
     {
         [Fact]
         public async Task Calendar_IsSupported_Fail_On_NetStandard() =>
-            await Assert.ThrowsAsync<NotImplementedInReferenceAssemblyException>(() => Task.FromResult(Calendar.IsSupported));
+            await Assert.ThrowsAsync<NotImplementedInReferenceAssemblyException>(() => Task.FromResult<bool>(Calendar.PlatformIsSupported));
```
Comment (nickrandolph) [inline Tests/Calendar_Tests.cs]: Change to 
async ()=> Calendar.PlatformSupported

File: Xamarin.Essentials/Calendar/Calendar.android.cs [hunk 5]
```diff
@@ -1,7 +1,56 @@
...
+
+            var cur = Platform.AppContext.ApplicationContext.ContentResolver.Query(calendarsUri, calendarsProjection.ToArray(), null, null, null);
+            var calendars = new List<ICalendar>();
+            while (cur.MoveToNext())
+            {
+                calendars.Add(new DeviceCalendar()
+                {
+                    Id = cur.GetString(calendarsProjection.IndexOf(CalendarContract.Calendars.InterfaceConsts.Id)),
+                    Name = cur.GetString(calendarsProjection.IndexOf(CalendarContract.Calendars.InterfaceConsts.CalendarDisplayName)),
+                    IsReadOnly = IsCalendarReadOnly(cur.GetInt(calendarsProjection.IndexOf(CalendarContract.Calendars.InterfaceConsts.CalendarAccessLevel)))
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/Calendar.android.cs]: Parse to CalendarAccess enum and pass the value into the IsCalendarReadOnly method

File: Xamarin.Essentials/Calendar/Calendar.ios.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+            foreach (var t in calendars)
+            {
+                calendarList.Add(new DeviceCalendar
+                {
+                    Id = t.CalendarIdentifier,
+                    Name = t.Title,
+                    IsReadOnly = !t.AllowsContentModifications
+                });
+                /*
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/Calendar.ios.cs]: remove comments

File: Xamarin.Essentials/Calendar/Calendar.ios.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                 var mySavedEvent = CalendarRequest.Instance.EventFromIdentifier(t.CalendarIdentifier);
+                */
+            }
+            return calendarList.AsReadOnly();
+        }
+
+        public static async Task PlatformRequestCalendarReadAccess()
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarRead);
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/Calendar.ios.cs]: Remove call to CalendarWrite

File: Xamarin.Essentials/Calendar/Calendar.ios.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public static async Task PlatformRequestCalendarWriteAccess()
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarRead);
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/Calendar.ios.cs]: Remove call to CalendarRead

File: Xamarin.Essentials/Calendar/Calendar.shared.cs [hunk 9]
```diff
@@ -1,6 +1,22 @@
-﻿namespace Xamarin.Essentials
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Xamarin.Essentials
 {
     public static partial class Calendar
     {
+        // public static List<Calendar> CalendarList { get; set; }
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/Calendar.shared.cs]: Remove comment

File: Xamarin.Essentials/Calendar/CalendarRequest.ios.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (nickrandolph) [inline Xamarin.Essentials/Calendar/CalendarRequest.ios.cs]: Check that these are all required

File: Xamarin.Essentials/GlobalSuppressions.cs [hunk 11]
```diff
@@ -0,0 +1,6 @@
+﻿// This file is used by Code Analysis to maintain SuppressMessage
```
Comment (nickrandolph) [inline Xamarin.Essentials/GlobalSuppressions.cs]: Remove this file

File: Xamarin.Essentials/GlobalSuppressions2.cs [hunk 12]
```diff
@@ -0,0 +1,6 @@
+﻿// This file is used by Code Analysis to maintain SuppressMessage
```
Comment (nickrandolph) [inline Xamarin.Essentials/GlobalSuppressions2.cs]: Remove this file

File: Xamarin.Essentials/Permissions/Permissions.ios.cs [hunk 13]
```diff
@@ -106,5 +145,41 @@ void LocationAuthCallback(object sender, CLAuthorizationChangedEventArgs e)
...
+            (bool granted, NSError e) =>
+            {
+                if (granted)
+                {
+                }
+                else
+                {
+                    throw new PermissionException($"{e} was not granted.");
+                }
+            });
```
Comment (nickrandolph) [inline Xamarin.Essentials/Permissions/Permissions.ios.cs:157]: Need to complete the task completion source

File: Xamarin.Essentials/Types/DeviceCalendar.shared.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    [Preserve(AllMembers = true)]
+    public class DeviceCalendar : ICalendar
+    {
+        public string Id { get; set; }
+
+        public string Name { get; set; }
+
+        public bool IsReadOnly { get; set; }
+
+        // public List<Event> EventList { get; set; }
```
Comment (nickrandolph) [inline Xamarin.Essentials/Types/DeviceCalendar.shared.cs]: Remove this property

File: Xamarin.Essentials/Types/DeviceCalendar.shared.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        public List<int> SetPositions { get; set; }
+    }
+
+    public class Attendee
+    {
+        public string Name { get; set; }
+
+        public string Email { get; set; }
+
+        // public IAttendeeDetails AttendeeDetails => DeviceSpecificAttendeeDetails;
```
Comment (nickrandolph) [inline Xamarin.Essentials/Types/DeviceCalendar.shared.cs]: Remove this property for the moment

**Review by nickrandolph — CHANGES_REQUESTED** [review]
See inline comments

**Review by nickrandolph — APPROVED** [review]

File: Samples/Samples.iOS/Info.plist (hunk 1 above)
Comment (ScottBTR) [reply Samples/Samples.iOS/Info.plist]: Changes amended in newest commit

**Pull request merged**